opensearch/src/error.rs

```diff
@@ -32,125 +32,75 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-use crate::http::{transport::BuildError, StatusCode};
-use std::{error, fmt, io};
+
+use crate::{
+    cert::CertificateError,
+    http::{transport, StatusCode},
+};
+
+pub(crate) type BoxError<'a> = Box<dyn std::error::Error + Send + Sync + 'a>;
 
 /// An error with the client.
 ///
 /// Errors that can occur include IO and parsing errors, as well as specific
-/// errors from Elasticsearch and internal errors from the client.
-#[derive(Debug)]
-pub struct Error {
-    kind: Kind,
+/// errors from OpenSearch and internal errors from the client.
+#[derive(Debug, thiserror::Error)]
+#[error(transparent)]
+pub struct Error(Kind);
+
+impl<E> From<E> for Error
+where
+    Kind: From<E>,
+{
+    fn from(error: E) -> Self {
+        Self(Kind::from(error))
+    }
 }
 
-#[derive(Debug)]
-enum Kind {
-    /// An error building the client
-    Build(BuildError),
-
-    /// A general error from this library
-    Lib(String),
-
-    /// HTTP library error
-    Http(reqwest::Error),
-
-    /// IO error
-    Io(io::Error),
-
-    /// JSON error
-    Json(serde_json::error::Error),
-}
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum Kind {
+    #[error("transport builder error: {0}")]
+    TransportBuilder(#[from] transport::BuildError),
 
-impl From<io::Error> for Error {
-    fn from(err: io::Error) -> Error {
-        Error {
-            kind: Kind::Io(err),
-        }
-    }
-}
+    #[error("certificate error: {0}")]
+    Certificate(#[from] CertificateError),
 
-impl From<reqwest::Error> for Error {
-    fn from(err: reqwest::Error) -> Error {
-        Error {
-            kind: Kind::Http(err),
-        }
-    }
-}
+    #[error("reqwest error: {0}")]
+    Reqwest(#[from] reqwest::Error),
```

Comment:
Ah I see, we use it at build time, may be `ClientBuilder` would be better one?

Reply:
We already expose `reqwest` types externally so that's not a new problem as such. But these variants aren't actually exposed other than via the descriptor text or debug formatting. But certainly happy to rename to something that makes more sense to specific cases